contrib/pkg/broker/server/server.go

```diff
@@ -52,14 +53,23 @@ func CreateHandler(c controller.Controller) http.Handler {
 	return router
 }
 
-// Start creates the HTTP handler based on an implementation of a
-// controller.Controller interface, and begins to listen on the specified port.
-func Start(serverPort int, c controller.Controller) {
-	glog.Infof("Starting server on %d\n", serverPort)
-	http.Handle("/", CreateHandler(c))
-	if err := http.ListenAndServe(":"+strconv.Itoa(serverPort), nil); err != nil {
-		panic(err)
+// Run creates the HTTP handler based on an implementation of a
+// controller.Controller interface, and begins to listen on the specified address.
+func Run(ctx context.Context, addr string, c controller.Controller) error {
+	glog.Infof("Starting server on %d\n", addr)
+	srv := http.Server{
+		Addr:    addr,
+		Handler: createHandler(c),
 	}
+	go func() {
+		<-ctx.Done()
+		c, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		defer cancel()
+		if srv.Shutdown(c) != nil {
+			srv.Close()
+		}
+	}()
```

Comment:
In the background wait for the done from the interrupt and shut everything down cleanly. nice.